### DISCUSSION TITLE
Let's talk about Appwrite Functions (Gen 2)

### DISCUSSION THREAD

**Discussion by Meldiron** [appwrite/appwrite#2985] [Ideas]
![GitHub Topics](https://user-images.githubusercontent.com/19310830/159428653-8e7c2c51-c5a5-4af4-822d-66fdd3733f9c.png)

Appwrite Functions got a **HUGE** upgrade in version 0.13, introducing a new execution model that improves execution times from as high as 2 seconds, to as low as 1 millisecond.

Our main goal of this version was to improve performance, and code quality. Now, let's focus on some **cool features**!

Here are some of my ideas:

- Allowing custom installation script, like `npm install`, to support custom installations. The main idea behind this is to support private repositories. Can anyone experienced confirm if this is a valid solution?
- Allowing custom build script, like `npm run build`. This would come in handy for runtimes like NodeJS when using TypeScript. Such a project needs to run `tsc` before starting to properly build into JavaScript. This would also allow CI/CD-like experience when the build process would be able to download the latest translation files from 3rd party on each build. Can anyone see a problem with this feature?
- Introduce many more outputs of functions to properly differentiate between user and developer data. Some examples are:
1. Output, status, exit code - for client
2. Stdout (log), Stderr (exceptions) - for developer
- Custom runtimes, so you can extend existing runtime and add new native libraries. This might come in handy when using libraries that require OS setup, such as FFMPEG or headless browsers

---

Let's Chat! How can we improve Appwrite Functions? What do you think about my idea? Can you see any problems or do you have an idea how to improve it even more? Can you think of a new idea we could implement?

**Comment by Meldiron** [comment]
Adding a `.appwriteignore` idea by @rdmchr https://github.com/appwrite/sdk-for-cli/issues/27

**Comment by Meldiron** [comment]
How would you like to manage environment variables from the Appwrite CLI setup?
- We can't put them in appwrite.json, people would tend to push them to version control (Git)
- I am not sure about taking them from some `.env` files, because these are ignored and someone might push function without having them, losing envs on the server
- Overall, it's scarry to be overwriting ENVs on the server

My initial idea is to:
When deploying a function, look inside the functions folder and search for `.env` file. Merge these env variables with ones on the server, and notify the user during deployment if there is a change to the env variable, with option Y/n (yes/no) to proceed. If a variable is new, add it. If a variable is missing, keep it on a server, but don't download it.
What do you think? e😇

**Reply by rdmchr** [reply]
Yeah I think your idea is great.

**Comment by s3ppo** [comment]
https://github.com/appwrite/appwrite/issues/2994 :)

**Comment by stnguyen90** [comment]
For custom install and build scripts, it might be good to expose a hook or so during the lifecycle that allows devs to execute some sort of code.

💯 to capturing logging.

Custom runtimes is great, but I feel like it kind of conflicts with OpenRuntimes architecture. If I can bring my own image, there's less need for OpenRuntimes build and whatnot. This makes me think of Gitlab's CI/CD architecture vs Github's Actions. I don't like Github's Actions because it requires you to use their images when running the action. With Gitlab CI/CD, however, I have the flexibility of running with any image.

**Comment by radvansky-tomas** [comment]
https://github.com/appwrite/appwrite/issues/3207

**Comment by radvansky-tomas** [comment]
I would probably also add something similar to firebase CLI / Node Cloud functions...

Where you can have main functions project, and its up to you to "flag/export" functions you want to become "cloud functions"

**Comment by arcanechimp** [comment]
#3253

**Comment by tananga** [comment]
Why by default I can only create functions in 4 languages:
- Node
- PHP
- Python
- Ruby

Why I cant create Dart? Why I need to do some extra work to run Dart functions?

**Reply by stnguyen90** [reply]
Not all runtimes are enabled by default. There are ~30 when factoring in the different versions. That would clutter the runtimes drop-down. You can enable whichever ones you want by editing an environment variable. See the [docs](https://appwrite.io/docs/functions#supportedRuntimes).

**Reply by tananga** [reply]
I know that but for new users this is very bad behaviour. At least latest version should be always visible by default. Who cares if the dropdown is long? Its good that its long. It means that you support a lot of languages and not only 4.

**Comment by rafagazani** [comment]
>Allowing custom installation script, like npm install, to support custom installations. The main idea behind this is to support private  repositories. Can anyone experienced confirm if this is a valid solution?

Would this allow using python's pandas?
